mcp/mcp-annotations-spring/src/main/java/org/springframework/ai/mcp/annotation/spring/AbstractClientMcpHandlerRegistry.java

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.mcp.annotation.spring;
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import io.modelcontextprotocol.spec.McpSchema;
+import org.springaicommunity.mcp.annotation.McpElicitation;
+import org.springaicommunity.mcp.annotation.McpLogging;
+import org.springaicommunity.mcp.annotation.McpProgress;
+import org.springaicommunity.mcp.annotation.McpPromptListChanged;
+import org.springaicommunity.mcp.annotation.McpResourceListChanged;
+import org.springaicommunity.mcp.annotation.McpSampling;
+import org.springaicommunity.mcp.annotation.McpToolListChanged;
+
+import org.springframework.aop.framework.autoproxy.AutoProxyUtils;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Base class for sync and async ClientMcpHandlerRegistries. Not intended for public use.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @see ClientMcpAsyncHandlersRegistry
+ * @see ClientMcpSyncHandlersRegistry
+ */
+abstract class AbstractClientMcpHandlerRegistry implements BeanFactoryPostProcessor {
+
+	protected Map<String, McpSchema.ClientCapabilities> capabilitiesPerClient = new HashMap<>();
+
+	protected ConfigurableListableBeanFactory beanFactory;
+
+	protected final Set<String> allAnnotatedBeans = new HashSet<>();
+
+	static final Class<? extends Annotation>[] CLIENT_MCP_ANNOTATIONS = new Class[] { McpSampling.class,
+			McpElicitation.class, McpLogging.class, McpProgress.class, McpToolListChanged.class,
+			McpPromptListChanged.class, McpResourceListChanged.class };
+
+	static final McpSchema.ClientCapabilities EMPTY_CAPABILITIES = new McpSchema.ClientCapabilities(null, null, null,
+			null);
+
+	@Override
+	public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
+		this.beanFactory = beanFactory;
+		Map<String, List<String>> elicitationClientToAnnotatedBeans = new HashMap<>();
+		Map<String, List<String>> samplingClientToAnnotatedBeans = new HashMap<>();
+		for (var beanName : beanFactory.getBeanDefinitionNames()) {
+			if (!beanFactory.getBeanDefinition(beanName).isSingleton()) {
+				// Only process singleton beans, not scoped beans
+				continue;
+			}
+			var foundAnnotations = scan(AutoProxyUtils.determineTargetClass(beanFactory, beanName));
+			if (!foundAnnotations.isEmpty()) {
+				this.allAnnotatedBeans.add(beanName);
+			}
+			for (var foundAnnotation : foundAnnotations) {
```

Comment:
I guess this block can be nested in side the `if (!foundAnnotations.isEmpty()) {` above